sdk/python/feast/go_server.py

```diff
@@ -249,96 +215,69 @@ def get_online_features(
+        self._go_server_background_thread.join()
 
 
-# https://www.geeksforgeeks.org/python-different-ways-to-kill-a-thread/
-class GoServerBackgroundThread(threading.Thread):
+class GoServerMonitorThread(threading.Thread):
     def __init__(
         self,
         name: str,
         shared_connection: GoServerConnection,
-        go_server_started: threading.Event,
+        go_server_first_started: threading.Event,
     ):
         threading.Thread.__init__(self)
         self.name = name
         self._shared_connection = shared_connection
-        self._go_server_started = go_server_started
+        self._is_cancelled = False
+        self.daemon = True
+        self._go_server_started = go_server_first_started
 
     def run(self):
         # Target function of the thread class
+        _logger.debug("Started monitoring thread to keep go feature server alive")
         try:
-            while True:
-                self._go_server_started.clear()
+            while not self._is_cancelled:
 
                 # If we fail to connect to grpc stub, terminate subprocess and repeat
+                _logger.info("Connecting to subprocess")
                 if not self._shared_connection.connect():
+                    _logger.info("Failed to connect, killing and retrying")
                     self._shared_connection.kill_process()
                     continue
-                self._go_server_started.set()
-                while True:
+                else:
+                    _logger.debug("Go feature server started")
+                    self._go_server_started.set()
+                _logger.info("Status: %s", self._is_cancelled)
+                while not self._is_cancelled:
                     try:
                         # Making a blocking wait by setting timeout to a very long time so we don't waste cpu cycle
                         self._shared_connection.wait_for_process(3600)
                     except subprocess.TimeoutExpired:
                         pass
+                    _logger.info(
+                        "No longer waiting for process: %s, %s, %s",
+                        self._shared_connection._process.pid,
+                        self._shared_connection._process.returncode,
+                        self._shared_connection.is_process_alive(),
+                    )
                     if not self._shared_connection.is_process_alive():
                         break
         finally:
             # Main thread exits
             self._shared_connection.kill_process()
 
-    def stop_go_server(self):
-        thread_id = self._get_id()
-        res = ctypes.pythonapi.PyThreadState_SetAsyncExc(
-            thread_id, ctypes.py_object(SystemExit)
-        )
-        # TODO: Review that kill process here but run also has to stop
-        if res > 1:
-            ctypes.pythonapi.PyThreadState_SetAsyncExc(thread_id, 0)
-        else:
-            self._shared_connection.kill_process()
```

Comment:
Just curious, why was this the original implementation, feel like the current implementation to stop the thread is cleaner. Was there some type of problem with the new implementation that forced us do this messy asyncexc stuff and if so, how did we fix it?

Reply:
The first implementation was written during hackweek with a bunch of other moving pieces; not everything was optimal. post-hackweek we're trying to polish everything up.